File: Lab9_10_Search_algorithm/bucket_sort/bucket_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <limits>

using namespace std;

vector<double> bucketSort(vector<double>& arr) {
    if (arr.empty()) return arr;

    // Находим минимальное и максимальное значения
    double min_val = *min_element(arr.begin(), arr.end());
    double max_val = *max_element(arr.begin(), arr.end());

    // Создаем корзины
    int bucket_count = arr.size();
    vector<vector<double>> buckets(bucket_count);

    // Распределяем элементы по корзинам
    for (double num : arr) {
        int index = static_cast<int>((num - min_val) / (max_val - min_val) * (bucket_count - 1));
        buckets[index].push_back(num);
    }

    // Сортируем каждую корзину и объединяем результат
    vector<double> sorted_arr;
    for (auto& bucket : buckets) {
        sort(bucket.begin(), bucket.end());
        sorted_arr.insert(sorted_arr.end(), bucket.begin(), bucket.end());
    }

    return sorted_arr;
}

// Пример использования
int main() {
    vector<double> arr = {0.42, 0.32, 0.33, 0.52, 0.37, 0.47, 0.51};

    cout << "Start array: ";
    for (double num : arr) {
        cout << num << " ";
    }
    cout << endl;

    vector<double> sorted_arr = bucketSort(arr);

    cout << "Sorted array: ";
    for (double num : sorted_arr) {
        cout << num << " ";
    }
    cout << endl;

    return 0;
}

File: Lab9_10_Search_algorithm/bucket_sort/bucket_sort.java
import java.util.*;

public class Main {

    public static double[] bucketSort(double[] arr) {
        if (arr.length == 0) return arr;

        // Находим минимальное и максимальное значения
        double minVal = arr[0];
        double maxVal = arr[0];
        for (double num : arr) {
            if (num < minVal) minVal = num;
            if (num > maxVal) maxVal = num;
        }

        // Создаем корзины
        int bucketCount = arr.length;
        List<List<Double>> buckets = new ArrayList<>();
        for (int i = 0; i < bucketCount; i++) {
            buckets.add(new ArrayList<>());
        }

        // Распределяем элементы по корзинам
        for (double num : arr) {
            int index = (int) ((num - minVal) / (maxVal - minVal) * (bucketCount - 1));
            buckets.get(index).add(num);
        }

        // Сортируем каждую корзину и объединяем результат
        double[] sortedArr = new double[arr.length];
        int index = 0;
        for (List<Double> bucket : buckets) {
            Collections.sort(bucket);
            for (double num : bucket) {
                sortedArr[index++] = num;
            }
        }

        return sortedArr;
    }

    // Пример использования
    public static void main(String[] args) {
        double[] arr = {0.42, 0.32, 0.33, 0.52, 0.37, 0.47, 0.51};

        System.out.print("Исходный массив: ");
        for (double num : arr) {
            System.out.print(num + " ");
        }
        System.out.println();

        double[] sortedArr = bucketSort(arr);

        System.out.print("Отсортированный массив: ");
        for (double num : sortedArr) {
            System.out.print(num + " ");
        }
        System.out.println();
    }
}
